### ISSUE TITLE
mpfit.py

### ISSUE THREAD

**Issue by kdere** [USNavalResearchLaboratory/eispac#10]
It was not necessary to recode mpfit for Python.  The same algorithm is available in scipy.optimize.leastsq

hope this is helpful

Ken

**Comment by MJWeberg** [comment]
Hello Ken,
Good point! Unfortunately, while scipy.optimize.leastsq (or the newer least_squares function) is based on the same minimization algorithm, it does not have all of the same features as MPFIT. In particular, none of the scipy fitting routines allow you to tie the value of one parameter to another, a powerful feature that we use a fair bit with the multi-component fit templates. We also looked at some of the other Python fitting packages, such as lmfit and astropy.modeling, but they either have some operational quirks or are too computationally inefficient for our use at this time.

Thankfully, we did not have to redevelop anything. Although it is not popular within the Python community, mpfit.py has been around and updated since 2002. Additionally, we ran an array of test cases and confirmed that the Python version gives comparable results to the most up-to-date IDL version. Nevertheless, we will monitor the development of other fitting packages and plan to eventually add options for selecting the routine used by eispac.fit_spectra().

Respectfully,
Micah W.

**Comment by Richardjmorton** [comment]
Hi, do you really need tied parameters when fitting? Scipy optimise minimize offers a couple of options for bounded minimisation (although I do not think tying is available). Plus, you can provide a Jacobian that is calculated using automatic differentiation and combine with JIT compilation (both of which can lead to a speed up of ~x100 compared to the standard minimize speed - I used JAX for something similar). It might be useful considering the amount of spectra that need fitting.

**Comment by hpwarren** [comment]
Yes, there are a number of cases where tied parameters are used. For example, the centroid of Fe XII 195.179 is assumed to have a fixed offset from Fe XII 195.119. The widths of the Fe XI 188 doublet are another example. They are assumed to be the same, not just within a range. 

It seems to me that a spectral fitting package for solar physics needs to be developed. Things that are developed for astronomy are often very slow and don't scale to cases were you have tens of millions of profiles per observation.￼

**Comment by Richardjmorton** [comment]
Hmm, from those two examples I think it would be straightforward to avoid using tied parameters (if you wanted to move away from mpfit).

I was wondering a couple of things. a) how long it takes for you to fit a 256 by 256 raster scan with you current method? b) what you time-frame you consider as slow?

While trying out the JAX package I could fit a 256 by 256 scan in ~45 minutes. I am a bit of a novice with JAX, so there could be other ways to increase speed. Plus JAX has GPU/TPU support.
